src/compiler/checker.ts

```diff
@@ -8520,6 +8520,17 @@ namespace ts {
         function getContextuallyTypedParameterType(parameter: ParameterDeclaration): Type {
             const func = parameter.parent;
             if (isContextSensitiveFunctionOrObjectLiteralMethod(func)) {
+                if (isIife(func)) {
+                    const indexOfParameter = indexOf(func.parameters, parameter);
+                    const call = func.parent.parent as CallExpression;
+                    if (indexOfParameter < call.arguments.length) {
+                        const type = getTypeOfExpression(call.arguments[indexOfParameter]);
+                        if (type && parameter.dotDotDotToken) {

```

Comment:
For a rest parameter I think you need to collect all of the remaining arguments starting at that position and compute a union of their types. For example:

```typescript
((...a) => {
    // Type of a should be number | string
})(1, 2, "hello");
```